File: branch_sums/branch_sums.java
import java.util.ArrayList;
import java.util.List;

public class BinaryTree {
    int value;
    BinaryTree left;
    BinaryTree right;

    BinaryTree(int value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }

    public static void main(String[] args) {
        BinaryTree bst = new BinaryTree(10);
        bst.left = new BinaryTree(8);
        bst.left.left = new BinaryTree(7);
        bst.left.right = new BinaryTree(9);
        bst.right = new BinaryTree(12);
        bst.right.left = new BinaryTree(11);
        bst.right.right = new BinaryTree(13);

        List<Integer> result = BranchSums.branchSums(bst);
        System.out.println(result);
    }
}

class BranchSums {

    public static List<Integer> branchSums(BinaryTree root) {
        List<Integer> sums = new ArrayList<>();
        calculateBranchSums(root, 0, sums);
        return sums;
    }

    private static void calculateBranchSums(BinaryTree node, int runningSum, List<Integer> sums) {
        if (node == null) {
            return;
        }

        int newRunningSum = runningSum + node.value;

        if (node.left == null && node.right == null) {
            sums.add(newRunningSum);
            return;
        }

        calculateBranchSums(node.left, newRunningSum, sums);
        calculateBranchSums(node.right, newRunningSum, sums);
    }
}

File: branch_sums/branch_sums.go
package main

import (
	"fmt"
)

type BinaryTree struct {
	Value int
	Left  *BinaryTree
	Right *BinaryTree
}

func BranchSums(root *BinaryTree) []int {
	var sums []int
	calculateBranchSums(root, 0, &sums)
	return sums
}

func calculateBranchSums(node *BinaryTree, runningSum int, sums *[]int) {
	if node == nil {
		return
	}

	newRunningSum := runningSum + node.Value

	if node.Left == nil && node.Right == nil {
		*sums = append(*sums, newRunningSum)
		return
	}

	calculateBranchSums(node.Left, newRunningSum, sums)
	calculateBranchSums(node.Right, newRunningSum, sums)
}

func main() {
	bst := &BinaryTree{Value: 10}
	bst.Left = &BinaryTree{Value: 8}
	bst.Left.Left = &BinaryTree{Value: 7}
	bst.Left.Right = &BinaryTree{Value: 9}
	bst.Right = &BinaryTree{Value: 12}
	bst.Right.Left = &BinaryTree{Value: 11}
	bst.Right.Right = &BinaryTree{Value: 13}

	result := BranchSums(bst)
	fmt.Println(result)
}
